clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/ReservationReportCachingAgent.groovy

```diff
@@ -110,90 +117,105 @@ class ReservationReportCachingAgent implements CachingAgent {
-        }.each {
-          def osType = operatingSystemType(it.productDescription)
-          def reservation = reservations["${it.availabilityZone}:${osType.name}:${it.instanceType}"]
-          reservation.totalReserved.addAndGet(it.instanceCount)
-
-          if (osType.isVpc) {
-            reservation.accounts[credentials.name].reservedVpc.addAndGet(it.instanceCount)
-          } else {
-            reservation.accounts[credentials.name].reserved.addAndGet(it.instanceCount)
+    Observable
+      .from(accounts.sort { it.name })
+      .flatMap({ sortedAccountCredentials ->
+      Observable
+        .from(sortedAccountCredentials)
+        .flatMap({ NetflixAmazonCredentials credentials ->
+        credentials.regions.each { AmazonCredentials.AWSRegion region ->
+          log.info("Fetching reservation report for ${credentials.name}:${region.name}")
+
+          def amazonEC2 = amazonClientProvider.getAmazonEC2(credentials, region.name)
+          def reservedInstancesResult = amazonEC2.describeReservedInstances(new DescribeReservedInstancesRequest())
+          reservedInstancesResult.reservedInstances.findAll {
+            it.state.equalsIgnoreCase("active") &&
+              ["Heavy Utilization", "Partial Upfront", "All Upfront", "No Upfront"].contains(it.offeringType)
+          }.each {
+            def osType = operatingSystemType(it.productDescription)
+            def reservation = getReservation("${it.availabilityZone}:${osType.name}:${it.instanceType}")
+            reservation.totalReserved.addAndGet(it.instanceCount)
+
+            if (osType.isVpc) {
+              reservation.getAccount(credentials.name).reservedVpc.addAndGet(it.instanceCount)
+            } else {
+              reservation.getAccount(credentials.name).reserved.addAndGet(it.instanceCount)
+            }
           }
-        }
-
-        def describeInstancesRequest = new DescribeInstancesRequest()
-        def allowedStates = ["pending", "running", "shutting-down", "stopping", "stopped"] as Set<String>
-        while (true) {
-          def result = amazonEC2.describeInstances(describeInstancesRequest)
-          result.reservations.each {
-            it.getInstances().each {
-              if (!allowedStates.contains(it.state.name.toLowerCase())) {
-                return
-              }
-
-              def osTypeName = operatingSystemType(it.platform ? "Windows" : "Linux/UNIX").name
-              def reservation = reservations["${it.placement.availabilityZone}:${osTypeName}:${it.instanceType}"]
-              reservation.totalUsed.incrementAndGet()
 
-              if (it.vpcId) {
-                reservation.accounts[credentials.name].usedVpc.incrementAndGet()
-              } else {
-                reservation.accounts[credentials.name].used.incrementAndGet()
+          def describeInstancesRequest = new DescribeInstancesRequest()
+          def allowedStates = ["pending", "running", "shutting-down", "stopping", "stopped"] as Set<String>
```

Comment:
Didn't really seem to add much value pulling it out of the loop so I've opt'd to leave it.